PSL/SENSORS/VL531X.py

```diff
@@ -0,0 +1,451 @@
+    (0x4B, 0x09),
+    (0x4C, 0x05),
+    (0x4D, 0x04),
+    (0xFF, 0x00),
+    (0x44, 0x00),
+    (0x45, 0x20),
+    (0x47, 0x08),
+    (0x48, 0x28),
+    (0x67, 0x00),
+    (0x70, 0x04),
+    (0x71, 0x01),
+    (0x72, 0xFE),
+    (0x76, 0x00),
+    (0x77, 0x00),
+    (0xFF, 0x01),
+    (0x0D, 0x01),
+    (0xFF, 0x00),
+    (0x80, 0x01),
+    (0x01, 0xF8),
+    (0xFF, 0x01),
+    (0x8E, 0x01),
+    (0x00, 0x01),
+    (0xFF, 0x00),
+    (0x80, 0x00),
+)
+
+# I2C general - probably implemented in firmware; not our concern here
+#  # datasheet p18: first byte is address and read/write direction
+#  WRITE_ADDRESS = 0x52
+#  READ_ADDRESS = 0x53
+
+# abbreviations and acronyms
+# VCSEL - Vertical Cavity Surface Emitting Laser
+# SPAD - Single Photon Avalanche Diode
+# MSRC - Minimum Signal Rate Check
+class VL531X():
+    PLOTNAMES = ['Data']
+    ADDRESS = 0x29
+    name = 'Time-of-Flight Proximity Sensor'
+
+    def __init__(self, handler, **args):
+        self.name = 'Time-of-Flight Proximity Sensor'
+        self.ADDRESS = args.get('address', self.ADDRESS)
+        self.I2C = I2CSlave(handler, self.ADDRESS)
+        '''
+        try:
+            print ('switching baud to 400k')
+            self.I2C.configI2C(400e3)
+        except:
+            print ('FAILED TO CHANGE BAUD RATE')
+        '''
+        self.io_timeout_s = 10
+        # TODO: yet unused; move up?
+        self.signal_rate_limit = 0.25
+
+        self.init()
+
+    def init(self):
+        # Check identification registers for expected values
+        # VL53L0X datasheet section 3.2 / p21
+        for reg, val in _ID_REGISTERS:
+            res = self.I2C.read_bulk_byte(reg)
+            if (res != val):
+                print(res, val)
+                raise RuntimeError(
+                    "Unexpected ID register values"
+                )
+
+        # Initial configuration
+        # 1 .Access to sensor, based on Adafruit's and Polulu's code
+        # See https://github.com/adafruit/Adafruit_CircuitPython_VL53L0X.git
+        # and https://github.com/pololu/vl53l0x-arduino/blob/master/VL53L0X.cpp
+        # I2C standard mode
+        # TODO: What do these registers and values actually mean?
+        for reg_and_val in (
+            (0x88, 0x00),
+            (0x80, 0x01),
+            (0xFF, 0x01),
+            (0x00, 0x00)
+        ):
+            self.I2C.write_bulk(reg_and_val)
+        self._stop_byte = self.I2C.read_bulk_byte(0x91)
+        # useful for debugging
+        # print("stop byte", self._stop_byte)
+        # TODO: What do these registers and values actually mean?
+        for reg_and_val in (
+            (0x00, 0x01),
+            (0xFF, 0x00),
+            (0x80, 0x00)
+        ):
+            self.I2C.write_bulk(reg_and_val)
+        # disable SIGNAL_RATE_MSRC (bit 1) and SIGNAL_RATE_PRE_RANGE (bit 4)
+        # limit checks
+        # TODO: What is MSRC config control?
+        config_control = self.I2C.read_bulk_byte(_MSRC_CONFIG_CONTROL) | 0x12
```

Comment:
What is 0x12? Avoid unnamed literals. The same goes for several lines following this one. If the purpose of a literal is unknown, mark it as TODO in a comment.

Reply:
Unfortunately, this is all proprietary. I couldn't figure out anything from the sources I ported this code from; see https://github.com/adafruit/Adafruit_CircuitPython_VL53L0X/blob/master/adafruit_vl53l0x.py which I referenced in this file.

Removing parts causes the sensor either not to work or to give wrong results. :/

Reply:
I see. Looking at this again, I now realize that `0x12 == 2**1 | 2**4`, where 1 and 4 are the bits mentioned in the commend above this line. So in this case, we should create two variables `disable_signal_rate_msrc = 0x2` and `disable_signal_rate_pre_range = 0x10`, and then replace 0x12 with `disable_signal_rate_msrc | disable_signal_rate_pre_range`.

Of course, for other literals there are no such explanations in the comments. In that case I suppose we have no choice but to accept the literals.